File: Project/Simulator/Tile.java
package Simulator;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import Simulator.Organisms.Organism;
import Utils.RNG;
import Utils.DynamicDirections;

public class Tile {
    private Tile[] directions;
    private List<Organism> organisms = new ArrayList<>();
    public final int index;

    public Tile(int index, int directionCount) {
        this.index = index;
        directions = new Tile[directionCount];
    }

    @Override
    public String toString() {
        return organisms.isEmpty() ? "" : organisms.get(0).getSymbol();
    }

    public void clear() {
        organisms.clear();
    }

    public void setLink(DynamicDirections direction, Tile tile) {
        directions[direction.ordinal()] = tile;
    }

    public Organism getOrganism() {
        return organisms.isEmpty() ? null : organisms.get(0);
    }

    public void placeOrganism(Organism organism) {
        organisms.add(organism);
    }

    public void removeOrganism(Organism organism) {
        organisms.remove(organism);
    }

    public Tile getRandomFreeNeighbour() {
        List<Tile> freeNeighbours = Arrays.stream(directions)
                .filter(tile -> tile != null && tile.isFree())
                .collect(Collectors.toList());
        if (freeNeighbours.isEmpty())
            return null;
        return freeNeighbours.get(RNG.roll(0, freeNeighbours.size()));
    }

    public Tile getNeighbour(DynamicDirections direction) {
        return directions[direction.ordinal()];
    }

    public Tile getRandomNeighbour() {
        List<Tile> neighbours = Arrays.stream(directions)
                .filter(tile -> tile != null)
                .collect(Collectors.toList());
        if (neighbours.isEmpty())
            return null;
        return neighbours.get(RNG.roll(0, neighbours.size()));
    }

    public DynamicDirections getRandomDirection() {
        List<DynamicDirections> availableDirections = Arrays.stream(directions)
                .filter(tile -> tile != null)
                .map(tile -> DynamicDirections.values()[Arrays.asList(directions).indexOf(tile)])
                .collect(Collectors.toList());
        if (availableDirections.isEmpty())
            return DynamicDirections.get("SELF");
        return availableDirections.get(RNG.roll(0, availableDirections.size()));
    }

    public boolean isFree() {
        return organisms.isEmpty();
    }

    public List<Tile> getNeighbours() {
        return Arrays.asList(directions);
    }

    public List<Tile> getOccupiedNeighbours() {
        return Arrays.stream(directions)
                .filter(tile -> tile != null && !tile.isFree())
                .collect(Collectors.toList());
    }

    public long getOrganismCount() {
        return organisms.size();
    }
}

File: Project/Simulator/Tile.py
import random
from typing import List

from .Organisms.Organism import Organism


class Tile:
    def __init__(self, index, directionCount):
        self.INDEX = index
        self.__directions = [None]*directionCount
        self.__organisms = []
        self.__parent = None

    def __str__(self):
        return '🔳' if not self.__organisms else self.__organisms[0].getSymbol()

    def clear(self):
        self.__organisms.clear()

    def getParent(self):
        return self.__parent

    def setParent(self, parent):
        self.__parent = parent

    def setLink(self, direction, tile):
        if (direction == None):
            return
        self.__directions[direction.ordinal()] = tile

    def getOrganism(self):
        return self.__organisms[0] if len(self.__organisms) > 0 else None

    def placeOrganism(self, organism: Organism):
        self.__organisms.append(organism)

    def removeOrganism(self, organism: Organism):
        self.__organisms.remove(organism)

    def getRandomFreeNeighbour(self):
        freeNeighbours = [
            tile for tile in self.__directions if tile != None and tile.isFree()]
        if len(freeNeighbours) == 0:
            return None
        return freeNeighbours[random.randint(0, len(freeNeighbours) - 1)]

    def getNeighbour(self, direction):
        return self.__directions[direction.ordinal()]

    def getRandomNeighbour(self):
        neighbours = [tile for tile in self.__directions if tile != None]
        if len(neighbours) == 0:
            return None
        return neighbours[random.randint(0, len(neighbours) - 1)]

    def isFree(self):
        return len(self.__organisms) == 0

    def getNeighbours(self):
        neighbours = []
        for direction in self.__directions:
            if direction is not None:
                neighbours.append(direction)
        return neighbours

    def getOccupiedNeighbours(self):
        occupiedNeighbours = [
            tile for tile in self.__directions if tile != None and not tile.isFree()]
        return occupiedNeighbours

    def getOrganismCount(self):
        return len(self.__organisms)
